### PR TITLE
Refactored Image and Text combination to use placeholder tokens instead of simple concatination

### PR DESCRIPTION
**Pull request by lusxvr**

### CODE REVIEW

File: models/config.py [hunk 1]
```diff
@@ -19,15 +19,17 @@ class VLMConfig:
     lm_rms_eps: float = 1e-5
     lm_re_base: int = 100000
     lm_max_position_embeddings: int = 8192
-    lm_vocab_size: int = 49152
+    lm_base_vocab_size: int = 49152
```
Comment (andimarafioti) [inline models/config.py:22]: why do you need this? isn't it a parameter of the original tokenizer?

File: models/config.py [hunk 2]
```diff
@@ -19,15 +19,17 @@ class VLMConfig:
...
+    extra_token_amount: int = 3  # Number of extra tokens for the VLM (image start, image end, image token)
```
[the remark this answers is not in the document]
Comment (lusxvr) [reply models/config.py:23]: I thought so too, but I did not really find a way. Normally, I would have just used a dictionary to store the tokens, because then I can simply get its length via `len()`. But since the config dataclass has to be immutable, I need a default factory with the `field` to make it work. This however, does not have a `len()` method. What I tried first is to have a placeholder in the config for the `lm_vocab_size` and then compute it from the base vocab size and the additional tokens in the `post_init()` of the dataclass, but this does not work since in the loading of a pre-trained VLM we overwrite the config on instantiation so a placeholder throws an error. Long story short: I agree with you that this is not a good way to hardcode the number like this, but after spending a bit too much time on trying to fix this yesterday, I thought it is better to stop, take this non-ideal solution for now since its very easy and it works, and tackle it later again after the rest of the PR has been implemented.

**Comment by lusxvr** [comment]
image

First training runs with the new embedding handling are in!

Blue: baseline run with current state of the main branch
Red: new embedding handling, otherwise same config as the other run

We can see that the losses follow each other very closely. I think considering that the whole tokenization and image/text combination logic was overworked in the PR this is exactly what we are looking for. While there is still a small difference, I think this could simply be due to the fact that we add additional `<image_start>` and `<image_end>` tokens to the `input_ids`, so the model does not see 100% the same things as it did before the refactor.

I would be interested to hear your take on this @andimarafioti, do you think this is something we have to investigate further or a complete match of the loss curves will never be achieved with this span of refactoring anyways and it is therefore fine?

Additionally, we can see that our current accuracy measurement is really not a good metric, both the current main and feature branch perform similarly bad, maybe since I did not pay any attention to the actual value of the learning rate (as long as it was the same between the two runs). Nevertheless, I wanted to include it for completeness.

Additionally, we can see that this new method of combining the image and text embeddings results in a ~40% increase in tokens/sec, which is a very welcome improvement!

**Comment by lusxvr** [comment]
The new embedding handing logic is ready! We train nanoVLM-450M with it and it achieves the same accuracy as the main branch with a slightly higher loss (since we have added indicative tokens of the image position now this makes sense).

image

File: data/collators.py [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
 
 class VQACollator(object):  # Visual Question Answering Collator
-    def __init__(self, tokenizer, max_length):
+    def __init__(self, tokenizer, max_length, mp_image_token_length):
         self.tokenizer = tokenizer
         self.max_length = max_length
+        self.mp_image_token_length = mp_image_token_length
+
+        self.boi_token_str = tokenizer.boi_token 
+        self.eoi_token_str = tokenizer.eoi_token
```
Comment (andimarafioti) [inline data/collators.py:10]: I don't think you need a boi and eoi, just an image delimiter token.

File: models/vision_language_model.py [hunk 4]
```diff
@@ -63,29 +98,44 @@ def forward(self, input_ids, image, attention_mask=None, targets=None):
...
+        # Create sequence indices for replacement.
+        seq_offsets = torch.arange(image_token_len, device=input_ids.device).unsqueeze(0)
+        sequence_idx_fill = start_indices.unsqueeze(1) + seq_offsets
         
+        # Perform the replacement using advanced indexing.
+        initial_combined_embeds[batch_idx_fill, sequence_idx_fill] = image_embd.to(initial_combined_embeds.dtype)
+
+        current_total_seq_len = initial_combined_embeds.size(1)
+        batch_size = image_embd.size(0) # Or initial_combined_embeds.size(0)
+
```
Comment (andimarafioti) [inline models/vision_language_model.py:131]: this is the same as in the forward method, right? It seems like a big enough chunk of code to create a function, smth like _replace_img_tokens_with_embeds

File: train.py [hunk 5]
```diff
@@ -302,7 +303,7 @@ def train(train_cfg, vlm_cfg):
 
             loss.backward()
 
-            if (i + 1) % train_cfg.gradient_accumulation_steps == 0 or i + 1 == len(train_loader):
+            if is_update_step:
```
Comment (andimarafioti) [inline train.py:306]: nice

File: train.py [hunk 6]
```diff
@@ -330,7 +330,7 @@ def train(train_cfg, vlm_cfg):
             batch_loss = mean(dist_gather(batch_loss)) if is_dist() else batch_loss  
             tokens_per_second = sum(dist_gather(tokens_per_second)) if is_dist() else tokens_per_second  
 
-            if train_cfg.eval_in_epochs and global_step % train_cfg.eval_interval == 0: #and is_master():
+            if train_cfg.eval_in_epochs and global_step % train_cfg.eval_interval == 0 and is_update_step:
```
Comment (andimarafioti) [inline train.py:333]: nice, this was actually a bit annoying xD

File: data/collators.py (hunk 3 above)
Comment (lusxvr) [reply data/collators.py:10]: In the end with the current logic I could also get rid of the start/end/delimiter tokens in general, since it checks for the first image token anyways. Do you think we even need a delimiter? Does it help the model? Because we would not really need it otherwise

File: data/collators.py [hunk 7]
```diff
@@ -31,39 +38,46 @@ def __call__(self, batch):
...
+        labels[:, -1] = -100 # Last token has no target
 
-        # Determine if sequences were truncated
+        # Determine original lengths before padding/truncation to handle truncation cases
         original_lengths = [len(self.tokenizer.encode(seq)) for seq in input_sequences]
         
+        # Calculate the number of tokens for the special image prefix (BOI, IMGs, EOI)
+        # This encoding should not add other special tokens like BOS/EOS for length calculation.
+        special_image_prefix_str = f"{self.boi_token_str}{self.image_token_str * self.mp_image_token_length}{self.eoi_token_str}"
+        num_special_prefix_tokens = len(self.tokenizer.encode(special_image_prefix_str, add_special_tokens=False))
```
[the remark this answers is not in the document]
Comment (lusxvr) [reply data/collators.py:50]: Good idea! Could save some time here propably
